pyrasa/utils/aperiodic_utils.py

```diff
@@ -58,85 +60,105 @@ def _get_gof(psd: np.ndarray, psd_pred: np.ndarray, fit_func: str) -> pd.DataFra
-
-    curv_kwargs = {
+    curv_kwargs: dict = {
         'maxfev': 10_000,
         'ftol': 1e-5,
         'xtol': 1e-5,
         'gtol': 1e-5,
-    }
-
-    off_guess = [aperiodic_spectrum[0]] if fit_bounds is None else fit_bounds[0]
-    exp_guess = (
-        [np.abs(np.log10(aperiodic_spectrum[0] / aperiodic_spectrum[-1]) / np.log10(freq[-1] / freq[0]))]
-        if fit_bounds is None
-        else fit_bounds[1]
-    )
-
-    valid_slope_functions = ['fixed', 'knee']
-    assert fit_func in valid_slope_functions, f'The slope fitting function has to be in {valid_slope_functions}'
-
-    if fit_func == 'fixed':
-        fit_f = fixed_model
-
-        p, _ = curve_fit(fit_f, freq, np.log10(aperiodic_spectrum))
-
-        params = pd.DataFrame(
-            {
-                'Offset': p[0],
-                'Exponent': p[1],
-                'fit_type': 'fixed',
-            },
-            index=[0],
-        )
-        psd_pred = fit_f(freq, *p)
-
-    elif fit_func == 'knee':
-        fit_f = knee_model  # type: ignore
-        # curve_fit_specs
-        cumsum_psd = np.cumsum(aperiodic_spectrum)
-        half_pw_freq = freq[np.abs(cumsum_psd - (0.5 * cumsum_psd[-1])).argmin()]
-        # make the knee guess the point where we have half the power in the spectrum seems plausible to me
-        knee_guess = [half_pw_freq ** (exp_guess[0] + exp_guess[0])]
-        # convert knee freq to knee val which should be 2*exp_1 but this seems good enough
-        curv_kwargs['p0'] = np.array(off_guess + knee_guess + exp_guess + exp_guess)  # type: ignore
-        # print(curv_kwargs['p0'])
-        # make this optional
-        curv_kwargs['bounds'] = ((0, 0, 0, 0), (np.inf, np.inf, np.inf, np.inf))  # type: ignore
-        # knee value should always be positive at least intuitively
-        p, _ = curve_fit(fit_f, freq, np.log10(aperiodic_spectrum), **curv_kwargs)
-
-        params = pd.DataFrame(
-            {
-                'Offset': p[0] / scale_factor,
-                'Knee': p[1],
-                'Exponent_1': p[2],
-                'Exponent_2': p[3],
-                'Knee Frequency (Hz)': p[1] ** (1.0 / (2 * p[2] + p[3])),
-                'fit_type': 'knee',
-            },
-            index=[0],
+    },
+    semi_log: bool = True,
+) -> tuple[pd.DataFrame, pd.DataFrame]:
+    """get the slope of the aperiodic spectrum"""
+
+    if isinstance(fit_func, str):
+        off_guess = [aperiodic_spectrum[0]] if fit_bounds is None else fit_bounds[0]
+        exp_guess = (
+            [np.abs(np.log10(aperiodic_spectrum[0] / aperiodic_spectrum[-1]) / np.log10(freq[-1] / freq[0]))]
+            if fit_bounds is None
+            else fit_bounds[1]
         )
-        psd_pred = fit_f(freq, *p)
+        valid_slope_functions = ['fixed', 'knee']
+        assert fit_func in valid_slope_functions, f'The slope fitting function has to be in {valid_slope_functions}'
+
+        if fit_func == 'fixed':
+            fit_f = fixed_model
```

Comment:
I would factor this out into its own function. Then you can simply either set `fit_func` to that function or even have the user supply it to you not as a string but as a function.